### PR TITLE
First pass on creating a feature that works with the backend

### PR DESCRIPTION
**Pull request by karabatov**
Such a document would have been extremely useful to me when I had just started, which is why I decided to write it.

I don't think this part of the app will change significantly, so it should be reasonably future-proof.

This is only a first pass, I'm intending to add sections on testing in the future and some strategies when working with more complicated data like e.g. using `decodeArraySafely`, subkeys without a response, form requests etc. 

This first section should be useful for newcomers, and in updates I want to collect some common practices which other people may find useful without the need to look for a place in the source where something is implemented in a certain way.

My intention with this version is that it is merged and is available while I'll be updating it with further PRs in the future.

### CODE REVIEW

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: nit: you should use `json` after the 3 backticks to have JSON syntax highlighting of the snippet

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: nit: you should use `swift` after the 3 backticks to get Swift syntax highlighting of the snippet (and same with all snippets below)

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: I'm wondering, do we take the habit of defining these structs all at top level, or do we prefer declaring them as nested types?

```swift
struct CitiesResponse: Equatable, Decodable {
  struct City: Equatable, Decodable {
    let name: String
  }
  let cities: [City]
}
```

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: I'm confused, isn't `AuthenticatedAccessible`'s `access` method expecting 2 parameters? I was not sure what the `with request: Request` was for when I saw your description of the protocol in the previous section, and since now in this example it's not even used, I'm not even sure if it's a typo or something I missed?

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]:
```suggestion
Note that `citiesBusinessController` is provided as a parameter to the feedback function and doesn't require `self`.
```

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: I don't see that "later section" in the document about extracting without having to define a dedicated `struct CitiesResponse` data type though?

**Review by AliSoftware — COMMENTED** [review]
Very interesting document! I'll probably never have to implement that myself but it is clear and was interesting to learn how all of this is architectured 👍

**Review by horothesun — APPROVED** [review]
Great introduction! It'll save a ton of time during onboarding 👍

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Ah right, I keep forgetting about it :)

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: AFAIK generally it makes more sense to define them at the top level, as you keep passing around the deepest-nested types around (for rendering etc.), if they were defined as subtypes you'd have to have typealiases everywhere or it would be much more typing.

Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: I'll make a note about that.

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: This is a convenience for when there is no `Request` (which is *quite* often). I'll make a note about it.

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Yeah, about that 😅 
I think I'll allude to more being written a little more subtly (just “this will be added later”) to get a more coherent document.

**Comment by karabatov** [comment]
@AliSoftware I made some changes based on your comments :)

- [x] Snippet types specified
- [x] Note about defining structs on the top level
- [x] Note about `accessible.access` second parameter
- [x] Typo fixed
- [x] Removed references to “later sections”, I'll put them back when I actually write them

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: 👍

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Makes sense now 👍

**Review by AliSoftware — APPROVED** [review]
Looking great now 👍 

Maybe now that you removed the in-document paragraph about "later sections" you could now add a section at the very end of the document like this?

```md
## Future documentation

This is is only an initial pass on the topic. Feel free to help improve that documentation. Ideas of topics to address include:

 - Provide an example of a resource access with a non-`Void` `Request`
 - Explain how we can avoid creating a `Decodable` dedicated type to decode the request for simpler responses
 - ...
```

**Comment by karabatov** [comment]
@AliSoftware I have a local todo on what I want to add to it, I'll be doing that soonish, so I don't feel there's a need to put future plans on the doc itself.

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (TheAdamBorek) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: You don't need that wrapper. In the BackendResource you can put the response handler with `Parser.decode(Parser.parse(data, key: "cities")` inside.

I know you have written that for the sake of this exercise you choose to stick with it, however it's something I would mark on PR to have less types if possible.

Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: I say that in the previous paragraph :) I will make an example of how to invoke `Parser` with a subkey later.

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (TheAdamBorek) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: can you also put the function signature/definition? 
it just feels odd to not have it.

Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Yeah, this was extracted from a convoluted example, I think I can define a super simple feedback here and have its full signature ✨

Comment (dmcrodrigues) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Agree, we should have a function like `whenLoading(cities:)` which aligns with our conventions.

**Review by dmcrodrigues — APPROVED** [review]
This looks good @karabatov, one detail or two but great initiative!

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (RuiAAPeres) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: You mention the `Builder`, but no description regarding it. Could you put a small reference to it and mention the proposal `Builders as functions` #62.

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (viorel15) [inline Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Maybe it's worth introducing the two available concrete auth  types: KongAuh and ClincialAuth.
Also an important thing is how you construct your resource using the KongAuth.patientId.

Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: I thought this was a little out of scope (also exposing unnecessary internal details), and anyway auth methods are probably going to change going forward with the work that @jtomanik is doing.

**Review by viorel15 — COMMENTED** [review]
I think it's worth adding some paragraphs for testing each component. 
For example testing the business controller should contain testing encoding/decoding of requests/responses  models, any other request attributes and specific errors decoding if any.
Also for view models we have specific strategy for testing them

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (viorel15) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: At minimum I suggest in removing the ConcreteAuth since there is no such type. 
Otherwise this will confuse even more and the reader would need at least another document to build a feature that works with backend.

**Comment by karabatov** [comment]
@viorel15 I'm going to introduce the testing section in a different PR after this is merged, as it's not that small. I think it's easier to process and review in chunks, that's why I only included a simple example here.

File: Cookbook/Technical-Documents/BackendFeatureHowTo.md
Comment (karabatov) [reply Cookbook/Technical-Documents/BackendFeatureHowTo.md]: Ah, confusing the reader, now that makes sense. I'll see what I can do about it. I just didn't want to  complicate the doc too much.

**Comment by viorel15** [comment]
> @viorel15 I'm going to introduce the testing section in a different PR after this is merged, as it's not that small. I think it's easier to process and review in chunks, that's why I only included a simple example here.

Ok. it makes sense since this is not trivial.

**Review by jtomanik — APPROVED** [review]

**Comment by karabatov** [comment]
@viorel15 ✅ used `KongAuth` and mentioned `ClinicalAuth`.
@dmcrodrigues @TheAdamBorek ✅ added full signature for `Feedback`.
@RuiAAPeres ✅ briefly described a builder and linked to the proposal.

**Comment by karabatov** [comment]
Looks like most everyone is fine with it, so I'll merge tomorrow conditional on no additional feedback, and will start working on making a second pass:
- Testing
- Various JSON parsing approaches
- Resource access with requests
- ✨

**Pull request merged**
